Actualizar ejemplo reactive, componente para errores, retoques menores.

starg09 opened the pull request:
## Cambios Menores

* Menú de navegación ahora marca con negrita el link actual.
* Botón de apostar (en reactive) ahora se colorea y muestra puntero al pasar el mouse.
* Removí (en reactive) un atributo `clr___` que quedaron de una versión vieja del ejemplo (que aparentemente usaba una librería de componentes/estilado).

---

## Cambios Principales

* Separado el mensaje de error en campos a un componente aparte.
* Los errores por campos inválidos ya no se muestran hasta que el campo es tocado.
* El botón de apostar no apuesta si el formulario no es válido.
  - Si no es valido, toca todos los campos así se muestran los errores.
* Arreglo: el validador custom de fecha ya no falla para hoy.
  - Ahora compara solo las fechas, ignorando la hora de los objetos.
* El validador custom para fecha ya no falla si el campo es nulo/vacío.
  - Sigue metodología de los validadores estándar, si el campo es obligatorio existe `Validators.required`
* Si el formulario está ok, pero hay un problema en el objeto apuesta, esos errores se muestran por consola y no se setea un resultado nulo.
  - Puede ocurrir con el monto mínimo de apuesta:
    - Constante "mínimo apuesta" en 10.
    - Validación interna de Apuesta (usada en template) mide que sea mayor al mínimo.
    - Validación de `Validators.min(...)` mide que sea mayor **o igual** al mínimo pasado.
    - Por eso reactive usa "mínimo + 1", sacando ese 1 el formulario pasa pero la apuesta falla.

---

Como comentario general nomás:

Leyendo sobre el tema le vi mas sentido a los reactivos si uno tuviera un formulario más complejo/dinámico - por ejemplo, si quisiéramos poder agregar un número indeterminado de apuestas en distintas fechas, que corran en simultaneo. Y para una fecha dada, multiples apuestas a correr (pleno a 1, pleno a 5, segunda docena)

* En template podría quedar un poco engorroso, al menos en teoría.
* Con reactive, al quedar toda la lógica del lado del código se puede generar a demanda los campos necesarios, y hacer un `@for` para todas las iteraciones del formulario. Se podría incorporar el tercer tipo de elemento del modulo, `FormArray`.

Dicho eso, no se me ocurrió un caso realmente práctico para usar. Y para cosas sencillas, hay poca diferencia entre ambas (siendo menos rebuscado el modo por template).

starg09 committed:
Actualizar ejemplo reactivo, componente para errores.

starg09 committed:
Arreglar tests rotos en reactive

Un mensaje se veía distinto, y el manejo de errores cambia ligeramente. En un lugar se modificaba la apuesta, no el formulario.

starg09 committed:
Arreglar errores de linter

starg09 committed:
Corregir un error mas de linter

starg09 committed:
Remover mensaje de debugeo

fdodino commented on docs/reactive-forms.md:
```diff
@@ -81,57 +70,92 @@ Cada control necesita indicar el nombre del formControl
-  if (validator === 'required' && error) return `Debe ingresar ${field}`
-  if (validator === 'min' && error) return `Debe ingresar un valor mayor para ${field}`
-  return error?.[validator]?.message ?? undefined
+protected getMensajeError(validator: string) {
```
por qué `protected`? Me queda la duda

starg09 replied:
Creo que se me colocó automático, no veo ninguna razón para tenerlo (dudo que alguien vaya a llamar a "ValidationErrrorMessageComponent().getMensajeError() desde otro lado 😅)

starg09 replied:
Está ese cambio en angular 14 parece, no le termino de ver el sentido honestamente pero tal vez en algun caso aplica. Y simplemente no lo veo 🤷🏻

![image](https://github.com/uqbar-project/eg-apuestas-angular/assets/1235066/70205a43-ab68-41b2-af70-bae262a588c5)

fdodino commented on docs/reactive-forms.md:
```diff
@@ -81,57 +70,92 @@ Cada control necesita indicar el nombre del formControl
+
+Para determinar si la fecha de apuesta es válida, definimos un validador propio. En su nivel mínimo, este debe ser una función que cumpla con tres requisitos:
+* Debe recibir como parametro un objeto de tipo `AbstractControl`.
+* Si el campo es valido, la función debe retornar null
```
```suggestion
* Si el campo es válido, la función debe retornar null
```

fdodino commented on docs/reactive-forms.md:
```diff
@@ -81,57 +70,92 @@ Cada control necesita indicar el nombre del formControl
+  - Usualmente se devuelve un objeto, con una clave arbitraria seteada a algun booleano.
+  - Podemos aprovechar esto para incluir un mensaje de error a mostrar en nuestro template.
+
+Para no tener multiples funciones esparcidas en nuestros componentes, podemos agruparlas en una clase:
```
```suggestion
Para no tener múltiples funciones esparcidas en nuestros componentes, podemos agruparlas en una clase:
```

fdodino commented on docs/reactive-forms.md:
```diff
@@ -81,57 +70,92 @@ Cada control necesita indicar el nombre del formControl
 ## Consecuencias
 
-- Para evitar que el ejemplo se extienda mucho, no implementamos la combinación de dropdowns para el tipo de apuesta vs. el valor a apostar. Solo permitimos apostar a pleno
+- Para evitar que el ejemplo se extienda mucho, no implementamos la combinación de dropdowns para el tipo de apuesta vs. el valor a apostar. Solo permitimos apostar a pleno.
```
El punto va si sos consistente y ponés punto en cada viñeta, si no es mejor no ponerlo. En otras viñetas hay puntos para separar oraciones pero no al final

fdodino commented on src/app/apuestas-reactive/apuestas-reactive.component.ts:
```diff
@@ -16,46 +17,42 @@ export class ApuestasReactiveComponent {
-        Validators.required,
-      ]
-    ],
+    fecha: ['', [Validators.required, DateValidator.equalOrGreaterThanToday]],
+    monto: ['', [Validators.required, Validators.min(MONTO_MINIMO_PLENO + 1)]],
+    valorApostado: [1, [Validators.required]],
```
supongo que la diferencia es si tenés Prettier instalado o no por eso el salto de línea, ahora no me jode pero si vamos a laburar en un proyecto estaría bueno chequear que tengamos los mismos plugins para que un commit no cambie tanto

starg09 replied:
Tengo prettier instalado, al parecer queda así con la config que hay ahora en `.prettierrc.json`

fdodino commented on src/app/apuestas-reactive/apuestas-reactive.component.ts:
```diff
@@ -16,46 +17,42 @@ export class ApuestasReactiveComponent {
     // otra alternativa: https://docs.rxweb.io/getting-started
 
+    // antes que nada, validamos posibles erroresApuesta, y de existir los hacemos visibles "tocándolos".
+    if (this.apuestaForm.invalid) {
+      this.apuestaForm.markAllAsTouched()
+      return
+    }
```
👏🏼 👏🏼 👏🏼 

Cuando lo quise implementar no me di cuenta de este detalle y lo rollbackeé

fdodino commented on src/app/apuestas-reactive/apuestas-reactive.component.ts:
```diff
@@ -16,46 +17,42 @@ export class ApuestasReactiveComponent {
       this.apuesta,
       { ...this.apuestaForm.value },
-      { fecha: fecha ? dayjs(fecha).toDate() : undefined },
+      { fecha: campoFecha ?  dayjs(campoFecha).toDate() : undefined },
```
🏅

fdodino reviewed (approved):
¡Maravilloso! Una pinturita. Gracias por el laburo! Te dejé algunos detalles menores, pero quedó impresionante!

starg09 committed:
Fix: Remover protected en validation-error-message, correcciones menores

starg09 committed:
Correr prettier en apuestas-reactive.component.ts

starg09 committed:
Mejorar cobertura de componente

Hago método para acceder a los errores, no deja testear (especificamente espiar) sobre la propiedad "errores". Mejor no tocarla directo desde el componente, aparte.
